mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala

```diff
@@ -865,4 +934,114 @@ class Completions(
             else compareByRelevance(o1, o2)
       end compare
 
+  object MatchExtractor:
+    def unapply(path: List[Tree]) =
+      path match
+        // foo mat@@
+        case (sel @ Select(qualifier, name)) :: _
+            if "match".startsWith(name.toString()) && text.charAt(
+              completionPos.start - 1
+            ) == ' ' =>
+          Some(qualifier)
+        // foo match @@
+        case (c: CaseDef) :: (m: Match) :: _
+            if completionPos.query.startsWith("match") =>
+          Some(m.selector)
+        // foo ma@tch (no cases)
+        case (m @ Match(
+              _,
+              CaseDef(Literal(Constant(null)), _, _) :: Nil,
+            )) :: _ =>
+          Some(m.selector)
+        case _ => None
+  end MatchExtractor
+
+  object CaseExtractor:
+    def unapply(path: List[Tree]): Option[(Tree, Tree)] =
+      path match
+        // foo match
+        // case None => ()
+        // ca@@
+        case (id @ Ident(name)) :: Block(stats, expr) :: parent :: _
+            if "case"
+              .startsWith(name.toString()) && isLastMatch(
+              stats
+            ) && expr == id =>
+          val selector = stats.last.asInstanceOf[Match].selector
+          Some((selector, parent))
+        // List(Option(1)).collect {
+        //   case Some(value) => ()
+        //   ca@@
+        // }
+        case (ident @ Ident(name)) :: Block(
+              _,
+              expr,
+            ) :: (_: CaseDef) :: (m: Match) :: parent :: _
+            if ident == expr && "case"
+              .startsWith(name.toString()) =>
+          Some((m.selector, parent))
+        // foo match
+        //  ca@@
+        case (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        // List(foo).map { ca@@ }
+        case (ident @ Ident(name)) :: Block(stats, expr) :: (appl @ Apply(
+              fun,
+              args,
+            )) :: _
+            if stats.isEmpty && ident == expr && "case".startsWith(
+              name.toString()
+            ) =>
+          Some((EmptyTree, appl))
+
+        case _ => None
+  end CaseExtractor
+
+  object CasePatternExtractor:
+    def unapply(path: List[Tree]) =
+      path match
+        // case Som@@
+        case (ident @ Ident(
+              name
+            )) :: (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        // case abc @ Som@@
+        case (ident @ Ident(name)) :: Bind(
+              _,
+              _,
+            ) :: (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        // case abc @ @@
+        case Bind(_, _) :: (c: CaseDef) :: (m: Match) :: parent :: _ =>
+          Some((m.selector, parent))
+        case _ => None
+
+  end CasePatternExtractor
+
+  object TypedCasePatternExtractor:
+    def unapply(path: List[Tree]) =
+      path match
+        // case _: Som@@ =>
+        case (ident @ Ident(name)) :: Typed(
```

Comment:
```suggestion
        case (ident @ Ident(name)) :: _ ::(c:CaseDef)::(m:Match)::parent::_
```
Could we actually accept anything like this?  Is Bind and Typed the only options? Because I think we could quite reduce the number of cases this way

In that case we could merge the 4 cases into two:
```
        case (ident @ Ident(name)) :: (_ : Typed | _: Bind) ::(c:CaseDef)::(m:Match)::parent::_
```
I think

Reply:
Then we could't distinguish `case abc @ So@@` and `case abc: So@@` when for the first one we would like `Some(value)` and for the second one `Some[_]`